homeassistant/components/watergate/__init__.py

```diff
@@ -0,0 +1,82 @@
+"""The Watergate integration."""
+
+from __future__ import annotations
+
+import logging
+
+from watergate_local_api import WatergateLocalApiClient
+from watergate_local_api.models import WebhookEvent
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_IP_ADDRESS, CONF_NAME, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.network import get_url
+
+from .const import DOMAIN, VALVE_ENTITY_NAME, WEBHOOK_VALVE_TYPE
+from .coordinator import WatergateDataCoordinator
+from .entity import WatergateConfigEntry, WatergateData
+
+_LOGGER = logging.getLogger(__name__)
+
+
+PLATFORMS: list[Platform] = [
+    Platform.VALVE,
+]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: WatergateConfigEntry) -> bool:
+    """Set up Watergate from a config entry."""
+    sonic_name = entry.data[CONF_NAME]
+    sonic_address = entry.data[CONF_IP_ADDRESS]
+
+    _LOGGER.debug(
+        "Setting up watergate local api integration for device: %s (IP: %s)",
+        sonic_name,
+        sonic_address,
+    )
+
+    webhook_id = entry.entry_id
+
+    hass.components.webhook.async_unregister(webhook_id)
+    hass.components.webhook.async_register(
+        DOMAIN, "Watergate", webhook_id, handle_webhook
+    )
+
+    watergate_client = WatergateLocalApiClient(sonic_address)
+
+    _LOGGER.debug("Registered webhook: %s", webhook_id)
+
+    coordinator = WatergateDataCoordinator(hass, watergate_client)
+
+    await coordinator.async_refresh()
+
+    instance_url = get_url(hass)
+
+    await watergate_client.async_set_webhook_url(
+        instance_url + "/api/webhook/" + webhook_id
+    )
+
+    entry.runtime_data = WatergateData(watergate_client, coordinator, sonic_name, {})
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+async def handle_webhook(hass: HomeAssistant, webhook_id, request):
+    """Handle incoming webhook request."""
+    body = await request.json()
+
+    _LOGGER.debug("Received webhook: %s", body)
+    data = WebhookEvent.parse_webhook_event(body)
+
+    body_type = body.get("type")
+
+    if (entity := hass.config_entries.async_get_entry(webhook_id)) is not None:
+        if body_type == WEBHOOK_VALVE_TYPE:
+            entity.runtime_data.entities[VALVE_ENTITY_NAME].webhook_update(data.state)
```

Comment:
why is this necessary? You have one data dict for everything anyways, can't we call `async_set_updated_data` from the coordinator (which will inform all the entities)

Reply:
I guess it looks better now. Thanks for all those suggestions. I wrote integration once, 2 years ago, everything is different now :)